File: clients/android/app/src/main/java/tv/plurx/app/data/OpticalModels.kt
package tv.plurx.app.data

import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonElement

@Serializable
data class OpticalDriveStateDto(
    val state: String,
    val media_generation: String? = null,
    val disc_id: String? = null,
    val title_id: String? = null,
    val session_id: String? = null,
    val reason: String? = null,
)

@Serializable
data class OpticalRequirementDto(val id: String, val status: String, val detail: String)

@Serializable
data class OpticalDiscSummaryDto(
    val id: String,
    val media_generation: String,
    val format: String,
    val display_title: String? = null,
    val volume_label: String? = null,
    val suggested_title_id: String? = null,
) {
    val title: String get() = display_title ?: volume_label ?: "Inserted disc"
}

@Serializable
data class OpticalDriveDto(
    val id: String,
    val name: String,
    val owner_node_id: String,
    val enabled: Boolean,
    val state: OpticalDriveStateDto,
    val requirements: List<OpticalRequirementDto> = emptyList(),
    val disc: OpticalDiscSummaryDto? = null,
)

@Serializable
data class OpticalTitleSummaryDto(
    val id: String,
    val duration_ms: Long? = null,
    val angles: Int = 1,
    val matched_item_id: Long? = null,
    val match_kind: String? = null,
)

@Serializable
data class OpticalDriveDiscDto(
    val drive: OpticalDriveDto,
    val titles: List<OpticalTitleSummaryDto> = emptyList(),
)

@Serializable
data class OpticalDiscDto(
    val disc_id: String,
    val format: String,
    val volume_label: String? = null,
    val display_title: String? = null,
) {
    val title: String get() = display_title ?: volume_label ?: "Inserted disc"
}

@Serializable
data class OpticalTrackDto(
    val index: Int = -1,
    val codec: String,
    val channels: Int? = null,
    val language: String? = null,
    val title: String? = null,
    val default: Boolean = false,
    val forced: Boolean = false,
)

@Serializable
data class OpticalFactsDto(
    val video_codec: String? = null,
    val width: Int? = null,
    val height: Int? = null,
    val hdr: String? = null,
    val duration_ms: Long? = null,
    val audio_streams: List<OpticalTrackDto> = emptyList(),
    val subtitle_streams: List<OpticalTrackDto> = emptyList(),
)

@Serializable
data class OpticalTitleDto(
    val disc_id: String,
    val title_id: String,
    val angles: Int = 1,
    val facts: OpticalFactsDto,
    val duration_ms: Long? = null,
    val matched_item_id: Long? = null,
    val match_kind: String? = null,
)

@Serializable
data class OpticalChapterDto(
    val index: Int? = null,
    val start_ms: Long? = null,
    val end_ms: Long? = null,
)

@Serializable
data class OpticalProgressDto(
    val disc_id: String,
    val title_id: String,
    val angle: Int,
    val position_ms: Long,
    val duration_ms: Long? = null,
    val watched: Boolean,
    val updated_at_ms: Long? = null,
)

@Serializable
data class OpticalTitleDetailDto(
    val disc: OpticalDiscDto,
    val title: OpticalTitleDto,
    val chapters: List<OpticalChapterDto> = emptyList(),
    val progress: OpticalProgressDto? = null,
)

@Serializable
data class OpticalDecisionRequest(
    val expected_disc_id: String,
    val media_generation: String,
    val angle: Int = 1,
    val caps: DeviceCaps,
    val force: String? = null,
    val audio: Int? = null,
    val subtitle: Int? = null,
)

@Serializable
data class OpticalDecisionDto(
    val method: String,
    val play_url: String,
    val source: OpticalFactsDto? = null,
    val audio: List<OpticalTrackDto> = emptyList(),
    val subtitles: List<OpticalTrackDto> = emptyList(),
    val ladder: List<Rung> = emptyList(),
    val delivered_dynamic_range: String? = null,
)

@Serializable
data class OpticalSessionRequest(
    val expected_disc_id: String,
    val media_generation: String,
    val angle: Int = 1,
    val playback_id: String,
    val request_id: String,
    val start: Double = 0.0,
    val height: Int? = null,
    val audio: Int? = null,
    val subtitle_burn: Int? = null,
    val audio_offset_ms: Long? = null,
    val block_budget_secs: Double? = null,
    val caps: DeviceCaps,
)

@Serializable
data class OpticalProgressRequest(
    val drive_id: String,
    val media_generation: String,
    val session_id: String,
    val angle: Int = 1,
    val position_ms: Long,
    val duration_ms: Long? = null,
    val audio: JsonElement? = null,
    val subtitle: JsonElement? = null,
    val recorded_at_ms: Long? = null,
)

@Serializable
data class OpticalEjectRequest(
    val media_generation: String,
    val session_id: String? = null,
    val stop_active: Boolean = false,
)

File: clients/apple/Sources/OpticalModels.swift
import Foundation

struct OpticalDriveStateDTO: Codable, Hashable {
    let state: String
    var mediaGeneration: String?
    var discId: String?
    var titleId: String?
    var sessionId: String?
    var reason: String?
}

struct OpticalRequirementDTO: Codable, Hashable, Identifiable {
    let id: String
    let status: String
    let detail: String
}

struct OpticalDiscSummaryDTO: Codable, Hashable {
    let id: String
    let mediaGeneration: String
    let format: String
    var displayTitle: String?
    var volumeLabel: String?
    var suggestedTitleId: String?

    var title: String { displayTitle ?? volumeLabel ?? "Inserted disc" }
}

struct OpticalDriveDTO: Codable, Hashable, Identifiable {
    let id: String
    let name: String
    let ownerNodeId: String
    let enabled: Bool
    let state: OpticalDriveStateDTO
    var requirements: [OpticalRequirementDTO]
    var disc: OpticalDiscSummaryDTO?
}

struct OpticalTitleSummaryDTO: Codable, Hashable, Identifiable {
    let id: String
    var durationMs: Int?
    let angles: Int
    var matchedItemId: Int?
    var matchKind: String?
}

struct OpticalDriveDiscDTO: Codable, Hashable {
    let drive: OpticalDriveDTO
    var titles: [OpticalTitleSummaryDTO]
}

struct OpticalDiscDTO: Codable, Hashable {
    let discId: String
    let format: String
    var volumeLabel: String?
    var displayTitle: String?

    var title: String { displayTitle ?? volumeLabel ?? "Inserted disc" }
}

struct OpticalTrackDTO: Codable, Hashable, Identifiable {
    var index: Int?
    let codec: String
    var channels: Int?
    var language: String?
    var title: String?
    var `default`: Bool?
    var forced: Bool?

    var id: String { "\(index ?? -1):\(language ?? ""):\(title ?? ""):\(codec)" }
}

struct OpticalFactsDTO: Codable, Hashable {
    var videoCodec: String?
    var width: Int?
    var height: Int?
    var hdr: String?
    var durationMs: Int?
    var audioStreams: [OpticalTrackDTO]?
    var subtitleStreams: [OpticalTrackDTO]?
}

struct OpticalTitleDTO: Codable, Hashable {
    let discId: String
    let titleId: String
    let angles: Int
    var facts: OpticalFactsDTO
    var durationMs: Int?
    var matchedItemId: Int?
    var matchKind: String?
}

struct OpticalChapterDTO: Codable, Hashable, Identifiable {
    var index: Int?
    var startMs: Int?
    var endMs: Int?

    var id: String { "\(index ?? 0):\(startMs ?? 0)" }
}

struct OpticalProgressDTO: Codable, Hashable {
    let discId: String
    let titleId: String
    let angle: Int
    let positionMs: Int
    var durationMs: Int?
    let watched: Bool
    var updatedAtMs: Int?
}

struct OpticalTitleDetailDTO: Codable, Hashable {
    let disc: OpticalDiscDTO
    let title: OpticalTitleDTO
    var chapters: [OpticalChapterDTO]
    var progress: OpticalProgressDTO?
}

struct OpticalDecisionRequest: Codable {
    let expectedDiscId: String
    let mediaGeneration: String
    let angle: Int
    let caps: DeviceCaps
    var force: String?
    var audio: Int?
    var subtitle: Int?
}

struct OpticalDecisionDTO: Codable {
    let method: String
    let playUrl: String
    var source: OpticalFactsDTO?
    var audio: [OpticalTrackDTO]
    var subtitles: [OpticalTrackDTO]
    var ladder: [QualityRung]
    var deliveredDynamicRange: String?
}

extension OpticalDecisionDTO {
    func playerDecision(selection: PrePlaySelection) -> Decision {
        Decision(
            fileId: nil,
            method: method,
            playUrl: playUrl,
            delivery: Delivery(
                mode: "transcode",
                url: nil,
                sessionsUrl: playUrl,
                aac: nil,
                requiresHls: true,
                preserveDolbyVision: false,
                audio: selection.audioIndex
            ),
            reasons: ["Managed optical source"],
            transcodeAudio: true,
            preserveDolbyVision: false,
            source: source.map {
                SourceSummary(
                    container: nil,
                    videoCodec: $0.videoCodec,
                    videoProfile: nil,
                    width: $0.width,
                    height: $0.height,
                    bitDepth: nil,
                    hdr: $0.hdr,
                    hdrFormat: nil,
                    bitrate: nil,
                    durationMs: $0.durationMs
                )
            },
            audio: audio.compactMap { track in
                track.index.map {
                    AudioTrack(
                        index: $0,
                        codec: track.codec,
                        channels: track.channels,
                        language: track.language,
                        title: track.title,
                        default: track.default ?? false
                    )
                }
            },
            subtitles: subtitles.compactMap { track in
                track.index.map {
                    SubtitleTrack(
                        index: $0,
                        codec: track.codec,
                        language: track.language,
                        title: track.title,
                        default: track.default ?? false,
                        forced: track.forced ?? false,
                        text: false,
                        native: false,
                        overlay: nil
                    )
                }
            },
            selection: DecisionSelection(
                audioIndex: selection.audioIndex,
                subtitleIndex: selection.subtitleIndex.flatMap { $0 >= 0 ? $0 : nil },
                subtitleRequiresBurnIn: selection.subtitleIndex.map { $0 >= 0 },
                subtitleBurnInBlockedByHdr: false
            ),
            markers: nil,
            audioOffsetMs: 0,
            declaredOffsetMs: 0,
            ladder: ladder,
            deliveredDynamicRange: deliveredDynamicRange,
            deliveredDolbyVisionProfile: nil
        )
    }
}

struct OpticalSessionRequest: Codable {
    let expectedDiscId: String
    let mediaGeneration: String
    let angle: Int
    let playbackId: String
    let requestId: String
    let start: Double
    var height: Int?
    var audio: Int?
    var subtitleBurn: Int?
    var audioOffsetMs: Int?
    var blockBudgetSecs: Double?
    let caps: DeviceCaps
}

struct OpticalProgressRequest: Codable {
    let driveId: String
    let mediaGeneration: String
    let sessionId: String
    let angle: Int
    let positionMs: Int
    var durationMs: Int?
    var audio: Int?
    var subtitle: Int?
    let recordedAtMs: Int
}

struct OpticalEjectRequest: Codable {
    let mediaGeneration: String
    var sessionId: String?
    let stopActive: Bool
}
